### ISSUE TITLE
WishartTriL sampling is very slow.

### ISSUE THREAD

**Issue by gileshd** [tensorflow/probability#1608]
It seems that sampling from a `WishartTriL` distribution is much slower than might be expected.

Using the jax substrate on a CPU it takes ~1s to generate a single 3x3 sample (not including any jax'y compilation which might be going on).

I've provided a MWE of this based on the example in the `WishartTriL` docstring:
```python
from jax import numpy as jnp
from jax import random as jr
from tensorflow_probability.substrates import jax as tfp
tfd = tfp.distributions

# Initialize a single 3x3 Wishart with Cholesky factored scale matrix and 5
# degrees-of-freedom.(*)
df = 5
scale_tril = jnp.tril(jnp.eye(3)/df)
chol_scale = jnp.linalg.cholesky(scale_tril)  # Shape is [3, 3].
dist = tfd.WishartTriL(df=df, scale_tril=chol_scale)

key = jr.PRNGKey(0)
# Run once in case of jax compilation shenanigans
W = dist.sample(seed=key)
W = dist.sample(sample_shape=10,seed=key)
W = dist.sample(sample_shape=100,seed=key)

%timeit dist.sample(seed=key).block_until_ready()
# ~1.0s

%timeit dist.sample(sample_shape=10,seed=key).block_until_ready()
# ~1.3s

%timeit dist.sample(sample_shape=100,seed=key).block_until_ready()
# ~2.6s
```
In contrast scipy generates a sample in ~160 $\mu s$:
```python
from scipy.stats import wishart
from jax import numpy as jnp

df = 5
scale = jnp.tril(jnp.eye(3)/df)
%timeit wishart.rvs(df,scale)
```

Python = 3.10.4
`jax.__version__` = '0.3.4'
`tfp.__version__` = '0.17.0'

**Comment by ColCarroll** [comment]
Hey! Thanks for the comment and for trying out the JAX substrate. I ran this on colab, and got somewhat faster times and better scaling than you (450ms - 500ms - 550ms, respectively). That's still much slower than scipy, though!

If you'd like to run fast, `jax.jit` will help:

```
sample = jax.jit(dist.sample, static_argnums=0)
```

In this case, the first run will do the compilation, and then subsequent calls (with the same shape) will be quite fast:

```python
sample = jax.jit(dist.sample, static_argnums=0)

%time sample(seed=key)  # 716 µs
%timeit sample(seed=key)  # 6.03 µs

%time sample(10, key)  # 711 µs
%timeit sample(10, key)  # 8.1 µs

%time sample(100, key)  # 1.49 ms
%timeit sample(100, key)  # 44.5 µs
```

**Comment by csuter** [comment]
Do you need to block_until_ready()?

**Comment by ColCarroll** [comment]
Sorry, I should have, but that only changes %time (I guess, since %timeit immediately calls the function again). 

In any case, I now get 1.5s - 1.9s - 1.8s for the initial tracing and sampling, but the same numbers once it is compiled.

**Comment by gileshd** [comment]
Ah ok! Good to know re `jit`.

I don't have a huge amount of experience with `tfp` and was presuming that some jitting was already happening under the hood and even then didn't expect this much of a difference. 

I've tested the jitted version and get similar results, I've seen some impressive speed ups with `jax.jit` but this one looks like a personal best with the jitted function running ~100,000x faster.

**Comment by ColCarroll** [comment]
Great! Yeah -- the policy has generally been to not jit _any_ code, for reasons having to do with TensorFlow, and since the debugging experience is at least different (if not better) when the code is not jitted.

**Comment by csuter** [comment]
maybe also need to split the seed? i remember something about failing to do
this causing no-ops in jax.
